compiler/src/dotty/tools/dotc/core/TypeErasure.scala

```diff
@@ -365,10 +365,18 @@ object TypeErasure {
   def erasedLub(tp1: Type, tp2: Type)(using Context): Type = {
     // After erasure, C | {Null, Nothing} is just C, if C is a reference type.
```

Comment:
I think this logic has gotten a bit messy and can be simplified. Why not:
```scala
if tp1.isRef(defn.NothingClass)
    || tp1.isRef(defn.NullClass) && tp2.derivesFrom(defn.ObjectClass))
then return tp2
```
And the same for `tp1`.